### PR TITLE
Add metadata property to configure Batching in Pulsar

### PR DESCRIPTION
**Pull request by saber-wang**
## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1671

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [√] Code compiles correctly
* [√] Created/updated tests
* [√] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2426

### CODE REVIEW

File: pubsub/pulsar/pulsar.go [hunk 1]
```diff
@@ -30,15 +30,17 @@ import (
...
-	tenant          = "tenant"
-	namespace       = "namespace"
-	persistent      = "persistent"
+	host                = "host"
+	consumerID          = "consumerID"
+	enableTLS           = "enableTLS"
+	deliverAt           = "deliverAt"
+	deliverAfter        = "deliverAfter"
+	disableBatching     = "disableBatching"
+	batchingMaxSize     = "batchingMaxSize"
```
Comment (shubham1172) [inline pubsub/pulsar/pulsar.go:39]: nit: order can be same in this struct and metadata.go

**Review by shubham1172 — DISMISSED** [review]
LGTM!

**Review by shubham1172 — APPROVED** [review]

File: pubsub/pulsar/pulsar_test.go [hunk 2]
```diff
@@ -37,6 +39,8 @@ func TestParsePulsarMetadata(t *testing.T) {
 	assert.Equal(t, true, meta.DisableBatching)
 	assert.Equal(t, defaultTenant, meta.Tenant)
 	assert.Equal(t, defaultNamespace, meta.Namespace)
+	assert.Equal(t, 100, meta.BatchingMaxSize)
```
Comment (shubham1172) [inline pubsub/pulsar/pulsar_test.go:42]: The tests are failing, the parsing logic does not seem to be working. Can you please take a look @saber-wang ?

**Review by shubham1172 — CHANGES_REQUESTED** [review]
Unit tests failing.

File: pubsub/pulsar/pulsar_test.go (hunk 2 above)
Comment (saber-wang) [reply pubsub/pulsar/pulsar_test.go:42]: @shubham1172  Execute make test and throw "make: no rule to make target 'test'. Stop" exception. How should I handle it?

Comment (shubham1172) [reply pubsub/pulsar/pulsar_test.go:42]: That is strange, the Makefile has the `test` command. What environment are on?

Comment (saber-wang) [reply pubsub/pulsar/pulsar_test.go]: @shubham1172 
 environment : win11 64bit, PowerShell 7.2.2,Make 4.3
I passed the test with "go test - V"

**Review by shubham1172 — DISMISSED** [review]

**Comment by shubham1172** [comment]
Thank you for the contribution, @saber-wang 🥳!

File: .gitignore [hunk 3]
```diff
@@ -3,4 +3,5 @@
 .vscode
 /vendor
 go.work
-.dccache
\ No newline at end of file
+.dccache
+/.vs
```
Comment (daixiang0) [inline .gitignore:7]: Theme unrelated changes, please remove.

File: pubsub/pulsar/pulsar.go [hunk 4]
```diff
@@ -92,7 +94,22 @@ func parsePulsarMetadata(meta pubsub.Metadata) (*pulsarMetadata, error) {
 		}
 		m.DisableBatching = disableBatching
 	}
-
+	m.BatchingMaxMessages = 1000
```
Comment (daixiang0) [inline pubsub/pulsar/pulsar.go:97]: Please add TODO and document those hardcode.

Comment (saber-wang) [reply pubsub/pulsar/pulsar.go]: I looked at the pulasr client's API and can remove these hard coded code

File: .history/pubsub/pulsar/pulsar_20220505155033.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
+/*
```
Comment (daixiang0) [inline .history/pubsub/pulsar/pulsar_20220505155033.go:1]: Please remove changes about `.history`.

File: .gitignore [hunk 6]
```diff
@@ -3,4 +3,3 @@
 .vscode
 /vendor
 go.work
-.dccache
```
Comment (shubham1172) [inline .gitignore:6]: Changes still there.

**Review by shubham1172 — APPROVED** [review]

File: pubsub/pulsar/pulsar.go [hunk 7]
```diff
@@ -92,7 +94,20 @@ func parsePulsarMetadata(meta pubsub.Metadata) (*pulsarMetadata, error) {
 		}
 		m.DisableBatching = disableBatching
 	}
-
+	if val, ok := meta.Properties[batchingMaxMessages]; ok {
```
Comment (shubham1172) [inline pubsub/pulsar/pulsar.go:97]: We don't have a default value here, and we rely on the upstream for it. So say if the default value of `bachingMaxMessages` get changed in the upstream, this will start using that on its own. 

However, if this value changes, our docs will be stale, since we are defining the default values there - https://github.com/dapr/docs/pull/2426/

Should we also hard code the defaults here to avoid that? Or instead of mentioning the default value in the docs, should we point to the upstream documentation? (I would prefer the former because it'll be less confusing for users). 

@saber-wang @daixiang0

Comment (saber-wang) [reply pubsub/pulsar/pulsar.go:97]: Hi, how can I format something like "5S" as "time. Duration"?

Comment (daixiang0) [reply pubsub/pulsar/pulsar.go:97]: Let's hardcode here and document it clearly.

Comment (shubham1172) [reply pubsub/pulsar/pulsar.go:97]: > And how can I format something like "5s" as "time. Duration"?

Sorry I did not understand this, can you explain more on what are you trying to do?

Comment (saber-wang) [reply pubsub/pulsar/pulsar.go:97]: @shubham1172  I have found `time.ParseDuration(val)` 😂.

**Review by shubham1172 — CHANGES_REQUESTED** [review]
As discussed, waiting for settings to be hardcoded.

File: pubsub/pulsar/pulsar.go [hunk 8]
```diff
@@ -317,3 +352,14 @@ func (p *Pulsar) formatTopic(topic string) string {
 	}
 	return fmt.Sprintf(topicFormat, persist, p.metadata.Tenant, p.metadata.Namespace, topic)
 }
+
+func formatDuration(durationString string) (time.Duration, error) {
```
Comment (saber-wang) [inline pubsub/pulsar/pulsar.go:356]: @shubham1172 Is there a better way?

File: pubsub/pulsar/pulsar.go [hunk 9]
```diff
@@ -50,6 +53,12 @@ const (
 	topicFormat      = "%s://%s/%s/%s"
 	persistentStr    = "persistent"
 	nonPersistentStr = "non-persistent"
+	// defaultBatchingMaxPublishDelay init default for maximum delay to batch messages
```
Comment (shubham1172) [inline pubsub/pulsar/pulsar.go:56]: Please fix linting issues.

File: pubsub/pulsar/pulsar.go [hunk 10]
```diff
@@ -50,6 +53,12 @@ const (
...
+	defaultBatchingMaxPublishDelay = 10 * time.Millisecond
```
Comment (shubham1172) [inline pubsub/pulsar/pulsar.go:57]: 1ms, same as upstream: https://pulsar.apache.org/docs/en/client-libraries-go/

File: pubsub/pulsar/pulsar.go (hunk 8 above)
Comment (shubham1172) [reply pubsub/pulsar/pulsar.go:356]: We can make it simpler by always expecting a number. 

The metadata can be `batchingMaxPublishDelayMs` to clearly indicate that it should be in milliseconds.

File: pubsub/pulsar/metadata.go [hunk 11]
```diff
@@ -13,13 +13,18 @@ limitations under the License.
...
-	Token           string `json:"token"`
+	Host                    string        `json:"host"`
+	ConsumerID              string        `json:"consumerID"`
+	EnableTLS               bool          `json:"enableTLS"`
+	DisableBatching         bool          `json:"disableBatching"`
+	Tenant                  string        `json:"tenant"`
+	Namespace               string        `json:"namespace"`
+	Persistent              bool          `json:"persistent"`
+	Token                   string        `json:"token"`
+	BatchingMaxPublishDelay time.Duration `json:"batchingMaxPublishDelay"`
```
Comment (shubham1172) [inline pubsub/pulsar/metadata.go:27]: Let's add the batching metadata right next to disable batching (since they are inter related). Same ordering for constants in `pulsar.go`.

File: pubsub/pulsar/pulsar.go (hunk 10 above)
Comment (saber-wang) [reply pubsub/pulsar/pulsar.go:57]: @shubham1172  Actually 10ms https://github.com/apache/pulsar-client-go/blob/5ee63303d43e8ee05f2c957a32c5526f29f54a36/pulsar/producer_impl.go#L38

**Review by berndverst — APPROVED** [review]

File: pubsub/pulsar/pulsar.go (hunk 8 above)
Comment (shubham1172) [reply pubsub/pulsar/pulsar.go:357]: @berndverst this comment was still pending.

Comment (saber-wang) [reply pubsub/pulsar/pulsar.go:357]: @shubham1172 Need to change to such a design?

Comment (shubham1172) [reply pubsub/pulsar/pulsar.go:357]: That would have been cleaner and simpler but Bernd has already approved/merged this PR. If we have enough consensus, we can create another PR on top to change that I guess.

Comment (saber-wang) [reply pubsub/pulsar/pulsar.go:357]: @shubham1172  I don't have a better idea. I can create a new rp to modify it

Comment (shubham1172) [reply pubsub/pulsar/pulsar.go:357]: Thanks @saber-wang. Let's see what Bernd/@daixiang0 have to say.
